### PR TITLE
feat: handle "file:/" and "file:///" for 'gf' command

### LINKED ISSUE #24032: Neovim sometimes includes a prefix to the filename for gf command when it shouldn't
### Problem

Neovim sometimes inclues a prefix to the filename for **gf** command when it shouldn't. It seems to happen only with an absolute path and not with a relative path.

<file1.txt>
file:/abs/path/to/tmp.txt

Action:
set isfname-=:
type "gf" when inside the file path

Result:
nvim tries to open "file:/abs/path/to/tmp.txt" instead of "/abs/path/to/tmp.txt"

### Steps to reproduce

<file1.txt>
file:/abs/path/to/tmp.txt

Action:
nvim --clean
set isfname-=:
type "gf" when inside the file path

### Expected behavior

nvim should open the file path without the leading "file:"

### Neovim version (nvim -v)

0.9.1 release

### Vim (not Nvim) behaves the same?

no, vim 9.0.1452

### Operating system/version

Red Hat Enterprise Linux Server release 7.9

### Terminal name/version

gnome-terminal

### $TERM environment variable

gnome-terminal

### Installation

appimage

### PR DESCRIPTION
**Pull request by jaehoonhwang**
Fixes #24032

Somehow, `file://` is a wrong one?  https://en.wikipedia.org/wiki/File_URI_scheme#How_many_slashes?

Changelog:
* Added more test cases and also have windows example. 
* Not handling `file://` as it is based on a host. 
* Handling hexcode on Windows; %A3 -> `;`

### CODE REVIEW

File: src/nvim/file_search.c [hunk 1]
```diff
@@ -1396,7 +1396,12 @@ char *find_file_in_path_option(char *ptr, size_t len, int options, int first, ch
     // filename on the first call.
     if (first == true) {
       if (path_with_url(*file_to_find)) {
-        file_name = xstrdup(*file_to_find);
+        if (strncmp(*file_to_find, "file", 4) == 0) {
```
Comment (justinmk) [inline src/nvim/file_search.c:1399]: `strncmp` avoids false positives, but it doesn't hurt to be extra clear for others reading the code.

```suggestion
        if (strncmp(*file_to_find, "file:/", 6) == 0) {
```

File: test/functional/core/path_spec.lua [hunk 2]
```diff
@@ -103,6 +103,14 @@ describe('file search', function()
     eq('filename_with_unicode_ααα', eval('expand("%:t")'))
   end)
 
+  it('find file name with prefix `file://` in line #24032', function()
```
Comment (justinmk) [inline test/functional/core/path_spec.lua:106]:
```suggestion
  it('"finds file://" URI on the local filesystem #24032', function()
```

File: src/nvim/file_search.c [hunk 3]
```diff
@@ -1396,7 +1396,12 @@ char *find_file_in_path_option(char *ptr, size_t len, int options, int first, ch
...
+          // convert "file://path/to/a" to "/path/to/a"
+          file_name = xstrdup(strchr(*file_to_find, '/'));
```
Comment (justinmk) [inline src/nvim/file_search.c:1401]: the test treats `file://foo` as `foo`, but the comment here implies that it's treated as `/foo` ?

File: test/functional/core/path_spec.lua [hunk 4]
```diff
@@ -103,6 +103,14 @@ describe('file search', function()
...
+    command('cd test/functional/fixtures')
+    insert('file://filename_with_unicode_ααα')
+    eq('', eval('expand("%")'))
+    feed('gf')
+    eq('filename_with_unicode_ααα', eval('expand("%:t")'))
```
Comment (justinmk) [inline test/functional/core/path_spec.lua:111]: need tests for `file:///` also. and other edge cases like `https://example.com/file:/foo.txt` . 

see also https://github.com/microsoft/vscode-uri/blob/53e4ca6263f2e4ddc35f5360c62bc1b1d30f27dd/src/test/uri.test.ts#L264-L268

Comment (jaehoonhwang) [reply test/functional/core/path_spec.lua]: Added the test and not handling `file://` (also added test cases for each of them)

PTAL

File: src/nvim/file_search.c [hunk 5]
```diff
@@ -1396,7 +1396,11 @@ char *find_file_in_path_option(char *ptr, size_t len, int options, int first, ch
     // filename on the first call.
     if (first == true) {
       if (path_with_url(*file_to_find)) {
-        file_name = xstrdup(*file_to_find);
+        if (strncmp(*file_to_find, "file:/", 6) == 0) {
+          file_name = xstrdup(handle_file_path_prefix(*file_to_find));
+        } else {
+          file_name = xstrdup(*file_to_find);
```
Comment (justinmk) [inline src/nvim/file_search.c:1402]: could be a ternary.

    file_name = ... ? ... : ...;

File: test/functional/core/path_spec.lua [hunk 6]
```diff
@@ -103,6 +105,36 @@ describe('file search', function()
...
+    test_gf("file:/c:/test/window/slash.txt", "/c:/test/window/slash.txt")
+    test_gf("file:///c:/test/window/3slash.txt", "///c:/test/window/3slash.txt")
+    test_gf("http://foo/bar/file:/with/url", "http://foo/bar/file:/with/url")
+
+    -- Don't handle `gf file://`
+    command('%delete')
+    insert("file://host/foo/bar/2slashes.txt")
+    command('norm! 0')
+    feed('gf')
+    neq("//host/foo/bar/2slashes.txt", eval('expand("%:p")'))
```
Comment (justinmk) [inline test/functional/core/path_spec.lua:129]: try to use positive confirmation (eq instead of neq), it results in must better failure messages and avoids false successes.

File: src/nvim/path.c [hunk 7]
```diff
@@ -2467,3 +2467,33 @@ void path_guess_exepath(const char *argv0, char *buf, size_t bufsize)
     xstrlcpy(buf, argv0, bufsize);
   }
 }
+
+/// Handles file path that starts with `file:[/|///]`
+///
+/// One and three slashes mean that file location is localhost/local.
+/// Example:
+/// - file:/path/to/file refers to /path/to/file
```
Comment (justinmk) [inline src/nvim/path.c:2475]: is there any system besides KDE which has the single-slash variant? 

is there a [vscode test](https://github.com/microsoft/vscode-uri/blob/53e4ca6263f2e4ddc35f5360c62bc1b1d30f27dd/src/test/uri.test.ts#L264-L268) which asserts this? 

please add a reference, I guess https://en.wikipedia.org/wiki/File_URI_scheme#Number_of_slash_characters is fine.

Comment (jaehoonhwang) [reply src/nvim/path.c:2475]: > is there any system besides KDE which has the single-slash variant?

Seems like this might be a good guide; http://eed3si9n.com/encoding-file-path-as-URI-reference/

Unix: u1/u3 `file:/` and `file:///`
Windows: u0/u3 `file:` and `file:///` (But, it seems like currently, neovim supports `

But, I don't know how to find if other system uses single-slash variant.  (googled a bit, but it leads to irrelevant stuff)

> is there a vscode test

This is trickier than I have thought before, as `:`(`0x3A`) and ` ` (`0x20`) are not allowed in file_path, Windows/DOS seems to convert those into hex code and Neovim needs to handle this case. 

For this one, it seems that `vim.uri_to_fname`([ref](https://github.com/neovim/neovim/blob/eadedfd9084c338a9162ef3706756bda443d6559/runtime/lua/vim/uri.lua#L102)) is available for file URI conversion, from lua side, but none seems to exist on C side? (I don't know if I'm understanding this correctly)

One thought is that it would be nice if c library side handles URIs nicely like lua side does. 


> please add a reference in this docstring, I guess https://en.wikipedia.org/wiki/File_URI_scheme#Number_of_slash_characters is fine.

Done.

**Comment by jaehoonhwang** [comment]
Oops, apparently, I don't know how to git smh

**Comment by justinmk** [comment]
Still interested in this, but probably want to move the logic to Lua, possibly using `vim.fs`.
